plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts

```diff
@@ -362,3 +352,129 @@ export class NunjucksWorkflowRunner implements WorkflowRunner {
+    help: 'Count of succesful task runs',
+    labelNames: ['template', 'invoker'],
+  });
+  const taskErrors = createCounterMetric({
+    name: 'scaffolder_task_error_count',
+    help: 'Count of failed task runs',
+    labelNames: ['template', 'invoker'],
+  });
+  const taskDuration = createHistogramMetric({
+    name: 'scaffolder_task_duration',
+    help: 'Duration of a task run',
+    labelNames: ['template', 'result'],
+  });
+  const stepSuccesses = createCounterMetric({
+    name: 'scaffolder_step_success_count',
+    help: 'Count of successful step runs',
+    labelNames: ['name'],
+  });
+  const stepErrors = createCounterMetric({
+    name: 'scaffolder_step_error_count',
+    help: 'Count of failed step runs',
+    labelNames: ['name'],
+  });
+  const stepDuration = createHistogramMetric({
+    name: 'scaffolder_step_duration',
+    help: 'Duration of a step runs',
+    labelNames: ['name', 'result'],
+  });
+
+  async function taskStart(task: TaskContext) {
+    await task.emitLog(`Starting up task with ${task.spec.steps.length} steps`);
+    const template = task.spec.templateInfo?.entity?.metadata.name || '';
+    const invoker = task.spec.user?.entity?.metadata.name || '';
+
+    const taskTimer = taskDuration.startTimer({
+      template,
+    });
+
+    async function skipDryRun(
+      step: TaskStep,
+      action: TemplateAction<JsonObject>,
+    ) {
+      task.emitLog(`Skipping because ${action.id} does not support dry-run`, {
+        stepId: step.id,
+        status: 'skipped',
+      });
+    }
+
+    function markSuccessful() {
+      taskSuccesses.inc({
+        template,
+        invoker,
+      });
+      taskTimer({ result: 'ok' });
+    }
+
+    async function markFailed(step: TaskStep, err: Error) {
+      await task.emitLog(String(err.stack), {
+        stepId: step.id,
+        status: 'failed',
+      });
+      taskErrors.inc({
+        template,
+        invoker,
+      });
+      taskTimer({ result: 'failed' });
+    }
+
+    return {
+      skipDryRun,
+      markSuccessful,
+      markFailed,
+    };
+  }
+
+  async function stepStart(task: TaskContext, step: TaskStep) {
+    await task.emitLog(`Beginning step ${step.name}`, {
+      stepId: step.id,
+      status: 'processing',
+    });
+
+    const stepTimer = stepDuration.startTimer({
+      name: step.name,
+    });
+
+    async function markSuccessful() {
+      await task.emitLog(`Finished step ${step.name}`, {
+        stepId: step.id,
+        status: 'completed',
+      });
+      stepSuccesses.inc({
+        name: step.name,
+      });
+      stepTimer({ result: 'ok' });
+    }
+
+    function markFailed() {
+      stepErrors.inc({
+        name: step.name,
+      });
+      stepTimer({ result: 'failed' });
+    }
+
+    async function skipFalsy() {
+      await task.emitLog(
+        `Skipping step ${step.id} because it's if condition was false`,
+        { stepId: step.id, status: 'skipped' },
+      );
```

Comment:
Is prometheus ok with leaving started timers just hanging? Do we need to cancel `stepTimer` here or should we do `stepTimer({ result: 'skipped' });` (which of course is a bit useless too because it'll usually be like zero milliseconds every time)?